src/Raven.Server/Documents/Handlers/IndexHandler.cs

```diff
@@ -1002,6 +1005,54 @@ await using (var ms = new MemoryStream())
             }
         }
 
+        [RavenAction("/databases/*/indexes/optimize", "POST", AuthorizationStatus.ValidUser, EndpointType.Write, DisableOnCpuCreditsExhaustion = true)]
+        public async Task OptimizeIndex()
+        {
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var index = Database.IndexStore.GetIndex(name);
+                if (index == null)
+                    throw new InvalidDataException($"Index '{name}' not found. Cannot perform optimize.");
+                var token = CreateOperationToken();
+                using (index.DrainRunningQueries())
+                {
+                    var result = new CompactionResult(index.Name);
+                    var operationId = ServerStore.Operations.GetNextOperationId();
+                    
+                    var t = ServerStore.Operations.AddOperation(
+                        Database,
+                        "Optimizing index: " + index.Name,
+                        Operations.Operations.OperationType.LuceneOptimizeIndex,
+                        taskFactory: onProgress => Task.Run(() =>
+                        {
+                            try
+                            {
+                                using (token)
+                                using (Database.PreventFromUnloadingByIdleOperations())
+                                using (var indexCts = CancellationTokenSource.CreateLinkedTokenSource(token.Token, Database.DatabaseShutdown))
+                                using (index.RestartEnvironment(onBeforeEnvironmentDispose: () => index.Optimize(onProgress, result, indexCts.Token)))
```

Comment:
This way we ensure that we are currently running any indexing, right?

Reply:
I don't see a problem with restarting the environment (even if not required for this operation), just asking to be sure